api/include/opentelemetry/trace/propagation/b3_propagator.h

```diff
@@ -0,0 +1,276 @@
+    }
+    return SpanContext::GetInvalid();
+  }
+
+  static TraceId GenerateTraceIdFromString(nostd::string_view trace_id)
+  {
+    uint8_t buf[kTraceIdHexStrLength / 2];
+    uint8_t *b_ptr = buf;
+    GenerateBuffFromHexStrPad0(trace_id, sizeof(buf), buf);
+    return TraceId(buf);
+  }
+
+  static SpanId GenerateSpanIdFromString(nostd::string_view span_id)
+  {
+    uint8_t buf[kSpanIdHexStrLength / 2];
+    uint8_t *b_ptr = buf;
+    GenerateBuffFromHexStrPad0(span_id, sizeof(buf), buf);
+    return SpanId(buf);
+  }
+
+  static TraceFlags GenerateTraceFlagsFromString(nostd::string_view trace_flags)
+  {
+    if (trace_flags.length() != 1 || (trace_flags[0] != '1' && trace_flags[0] != 'd'))
+    {  // check for invalid length of flags and treat 'd' as sampled
+      return TraceFlags(0);
+    }
+    return TraceFlags(TraceFlags::kIsSampled);
+  }
+
+private:
+  // Converts hex numbers (string_view) into bytes stored in a buffer and pads buffer with 0.
+  static void GenerateBuffFromHexStrPad0(nostd::string_view hexStr, int bufSize, uint8_t *buf)
+  {  // we are doing this starting from "right" side for left-padding
+    nostd::string_view::size_type posInp = hexStr.length();
+    int posOut                           = bufSize;
+    while (posOut--)
+    {
+      int val = 0;
+      if (posInp)
+      {
+        int hexDigit2 = HexToInt(hexStr[--posInp]);  // low nibble
+        int hexDigit1 = 0;
+        if (posInp)
+        {
+          hexDigit1 = HexToInt(hexStr[--posInp]);
+        }
+        if (hexDigit1 < 0 || hexDigit2 < 0)
+        {  // malformed hex sequence. Fill entire buffer with zeroes.
+          for (int j = 0; j < bufSize; j++)
+          {
+            buf[j] = 0;
+          }
+          return;
+        }
+        val = hexDigit1 * 16 + hexDigit2;
+      }
+      buf[posOut] = val;
+    }
+  }
+
+  // Converts a single character to a corresponding integer (e.g. '1' to 1), return -1
+  // if the character is not a valid number in hex.
+  static uint8_t HexToInt(char c)
+  {
+    if (c >= '0' && c <= '9')
+    {
+      return (int)(c - '0');
+    }
+    else if (c >= 'a' && c <= 'f')
+    {
+      return (int)(c - 'a' + 10);
+    }
+    else if (c >= 'A' && c <= 'F')
+    {
+      return (int)(c - 'A' + 10);
+    }
+    else
+    {
+      return -1;
+    }
+  }
+
+  static SpanContext ExtractImpl(Getter getter, const T &carrier)
+  {
+    // all these are hex values
+    nostd::string_view trace_id;
+    nostd::string_view span_id;
+    nostd::string_view trace_flags;
+
+    // first let's try a single-header variant
+    auto singleB3Header = getter(carrier, kB3CombinedHeader);
+    if (!singleB3Header.empty())
+    {
+      // From: https://github.com/openzipkin/b3-propagation/blob/master/RATIONALE.md
+      // trace_id can be 16 or 32 chars
+      auto firstSep = singleB3Header.find('-');
+      trace_id      = singleB3Header.substr(0, firstSep);
```

Comment:
This will allow an invalid length trace id and span id to pass through, e.g. `a-b` will produce `000000000000000a` as the trace ID.